### PR TITLE
Fix BaseMaterial3D refracting objects located in front of the material

### PR DESCRIPTION
**Pull request by Calinou**
Depth comparison is now used to prevent refraction from occurring if the pixel being refracted is located in front of the object.

For pixels slightly behind the object, a `smoothstep()` curve is used to progressively increases refraction intensity as the distance between the object and the refraction increases. This avoids sudden discontinuities in the refraction.

~~`ref_normal.xx` is also used instead of `ref_normal.xy` to prevent higher refraction scales from offsetting the entire scene behind the refraction.~~
**Edit:** Reverted, see https://github.com/godotengine/godot/pull/93449#issuecomment-2311012034.

Thanks to @Meccanizer for providing the depth comparison code (which I adapted to use `smoothstep()`) :slightly_smiling_face: 

- This closes https://github.com/godotengine/godot-proposals/discussions/9580.

**Testing projects:**

- [test_refraction_depth.zip](https://github.com/user-attachments/files/15934675/test_refraction_depth.zip)
- https://github.com/godotengine/godot-demo-projects/tree/master/3d/material_testers

## TODO

- [ ] Find ways to reduce the slight artifacting around object edges (for objects that are present in front of the reflection). See if the transition around these object edges can be made smoother as well when we lack screen-space information.

Can be done in a separate PR (as this also occurs in `master`):

- [ ] Progressively reduce the refraction amount when the screen edges would be reached to prevent streaks of long pixels (due to off-screen areas being read). You can see the issue at the bottom of the screenshots with refraction enabled.
- [ ] Progressively reduce the refraction amount according to the alpha value set by proximity fade or distance fade. This would avoid "double image" effects in such situations, as using proximity/distance fade is quite common for water effects.

## Preview

Before | After *(this PR)*
-|-
Before | After

For reference, this is how it looks like with depth comparison but without the `smoothstep()` curve:

Depth no mix

And this is how it looks without refraction:

No refraction

### Before

*Look at the ice material behind the orange toon material.*

https://github.com/godotengine/godot/assets/180032/577275b2-0f42-4d94-9d0e-932eeb444636

### After

*Look at the ice material behind the orange toon material.*

https://github.com/godotengine/godot/assets/180032/5d81d962-ee0c-4b1a-9ae4-6cf1955a9f00

### CODE REVIEW

**Comment by Meccanizer** [comment]
Thanks @Calinou for helping to fix my hyper specific problem. On further testing of this, I realized the reason why the refractions looks all weird is due to the ref_nomal.xy  unnaturally pulling the screen_uv down based on the cameras distance. changing the line from ref_normal.xy to ref_normal.xx fixes this problem, it removes the weird warping as well as only leaves the refractions. You can now set the refractions value as high as you like and there will be no distortion on the screen uv.


ref_normal.xy, using high refraction value to show distortion.
image

ref_normal.xx, screen_uv is now always in correct position and doesn't change based on cameras position
image

Only other problem I found was that uniform refraction_texture_channel doesn't have a default value so when you convert the material to a shader and run the game, the refractions go away in game but stay in the editor until you change it in the inspector. Setting a default value like vec4(0.0,0.0,1.0,0.0) should be fine.

I also found an article on writing refraction code for water on Unity. This article goes through fixes distortions and patching holes that appear when using the depth texture to cull objects in front. I'm still just a beginner when it comes to shader code and can barely read or understand Unity shader code so if someone wants to go through this and add anything useful that would be awesome. https://catlikecoding.com/unity/tutorials/flow/looking-through-water/

**Comment by Calinou** [comment]
> Only other problem I found was that uniform refraction_texture_channel doesn't have a default value so when you convert the material to a shader and run the game, the refractions go away in game but stay in the editor until you change it in the inspector. Setting a default value like vec4(0.0,0.0,1.0,0.0) should be fine.

I noticed that too while testing. Setting a default value did the trick, thanks :slightly_smiling_face: 

I used `vec4(1.0, 0.0, 0.0, 0.0)` as this matches the Red channel default used throughout BaseMaterial3D. (Note that ORM mode ignores the metallic texture channel uniform default, as it's not even defined in ORM mode.)

File: scene/resources/material.cpp [hunk 1]
```diff
@@ -1598,13 +1598,20 @@ void fragment() {)";
...
 	float ref_amount = 1.0 - albedo.a * albedo_tex.a;
-	EMISSION += textureLod(screen_texture, ref_ofs, ROUGHNESS * 8.0).rgb * ref_amount * EXPOSURE;
+
+	float refraction_depth_tex = textureLod(depth_texture, ref_ofs, 0.0).r;
+	vec4 refraction_world_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, refraction_depth_tex, 1.0);
+	refraction_world_pos.xyz /= refraction_world_pos.w;
+
+	// If the depth buffer is lower then the model's Z position, use the refracted UV, otherwise use the normal screen UV.
+	// At low depth differences, decrease refraction intensity to avoid sudden discontinuities.
+	EMISSION += textureLod(screen_texture, mix(SCREEN_UV, ref_ofs, smoothstep(0.0, 1.0, VERTEX.z - refraction_world_pos.z)), ROUGHNESS * 8.0).rgb * albedo.rgb * ref_amount * EXPOSURE;
```
[the remark this answers is not in the document]
Comment (Calinou) [reply scene/resources/material.cpp:1614]: Indeed, I removed it and it looks the same as before.

File: scene/resources/material.cpp [hunk 2]
```diff
@@ -1621,13 +1621,20 @@ void fragment() {)";
 )";
 		}
 		if (flags[FLAG_UV1_USE_TRIPLANAR]) {
-			code += "	vec2 ref_ofs = SCREEN_UV - ref_normal.xy * dot(triplanar_texture(texture_refraction, uv1_power_normal, uv1_triplanar_pos), refraction_texture_channel) * refraction;\n";
+			code += "	vec2 ref_ofs = SCREEN_UV - ref_normal.xx * dot(triplanar_texture(texture_refraction, uv1_power_normal, uv1_triplanar_pos), refraction_texture_channel) * refraction;\n";
 		} else {
-			code += "	vec2 ref_ofs = SCREEN_UV - ref_normal.xy * dot(texture(texture_refraction, base_uv), refraction_texture_channel) * refraction;\n";
+			code += "	vec2 ref_ofs = SCREEN_UV - ref_normal.xx * dot(texture(texture_refraction, base_uv), refraction_texture_channel) * refraction;\n";
```
Comment (huwpascoe) [inline scene/resources/material.cpp:1626]: Should just be .xy unless a proper transform has been discovered...

**Comment by huwpascoe** [comment]
Here's some distance correction.
```glsl
float DISTANCE_FACTOR = 0.1;
float distance_scale = clamp(1.0 / (length(VERTEX.xyz) * DISTANCE_FACTOR + 1.0), 0.0, 1.0);

EMISSION += textureLod(screen_texture, mix(SCREEN_UV, ref_ofs, distance_scale * smoothstep(0.0, 1.0, VERTEX.z - refraction_world_pos.z)), ROUGHNESS * 8.0).rgb * ref_amount * EXPOSURE;
```

File: scene/resources/material.cpp [hunk 3]
```diff
@@ -1627,7 +1627,14 @@ void fragment() {)";
 		}
 		code += R"(
 	float ref_amount = 1.0 - albedo.a * albedo_tex.a;
-	EMISSION += textureLod(screen_texture, ref_ofs, ROUGHNESS * 8.0).rgb * ref_amount * EXPOSURE;
+
+	float refraction_depth_tex = textureLod(depth_texture, ref_ofs, 0.0).r;
+	vec4 refraction_world_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, refraction_depth_tex, 1.0);
```
Comment (clayjohn) [inline scene/resources/material.cpp:1632]:
```suggestion
	vec4 refraction_view_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, refraction_depth_tex, 1.0);
```
This is in view space, not world space. I know that it was wrong in the proximity code too, but we should take the chance to clean this up

File: scene/resources/material.cpp [hunk 4]
```diff
@@ -1627,7 +1627,14 @@ void fragment() {)";
...
+	refraction_world_pos.xyz /= refraction_world_pos.w;
```
Comment (clayjohn) [inline scene/resources/material.cpp:1633]:
```suggestion
	refraction_view_pos.xyz /= refraction_view_pos.w;
```

File: scene/resources/material.cpp [hunk 5]
```diff
@@ -1627,7 +1627,14 @@ void fragment() {)";
...
+
+	// If the depth buffer is lower then the model's Z position, use the refracted UV, otherwise use the normal screen UV.
+	// At low depth differences, decrease refraction intensity to avoid sudden discontinuities.
+	EMISSION += textureLod(screen_texture, mix(SCREEN_UV, ref_ofs, smoothstep(0.0, 1.0, VERTEX.z - refraction_world_pos.z)), ROUGHNESS * 8.0).rgb * ref_amount * EXPOSURE;
```
Comment (clayjohn) [inline scene/resources/material.cpp:1637]:
```suggestion
	EMISSION += textureLod(screen_texture, mix(SCREEN_UV, ref_ofs, smoothstep(0.0, 1.0, VERTEX.z - refraction_view_pos.z)), ROUGHNESS * 8.0).rgb * ref_amount * EXPOSURE;
```

File: scene/resources/material.cpp [hunk 6]
```diff
@@ -1649,10 +1656,10 @@ void fragment() {)";
...
 		code += R"(
 	// Proximity Fade: Enabled
-	float depth_tex = textureLod(depth_texture, SCREEN_UV, 0.0).r;
-	vec4 world_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, depth_tex, 1.0);
-	world_pos.xyz /= world_pos.w;
-	ALPHA *= clamp(1.0 - smoothstep(world_pos.z + proximity_fade_distance, world_pos.z, VERTEX.z), 0.0, 1.0);
+	float proximity_depth_tex = textureLod(depth_texture, SCREEN_UV, 0.0).r;
+	vec4 proximity_world_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, proximity_depth_tex, 1.0);
+	proximity_world_pos.xyz /= proximity_world_pos.w;
+	ALPHA *= clamp(1.0 - smoothstep(proximity_world_pos.z + proximity_fade_distance, proximity_world_pos.z, VERTEX.z), 0.0, 1.0);
```
Comment (clayjohn) [inline scene/resources/material.cpp:1662]:
```suggestion
	vec4 proximity_view_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, proximity_depth_tex, 1.0);
	proximity_view_pos.xyz /= proximity_view_pos.w;
	ALPHA *= clamp(1.0 - smoothstep(proximity_view_pos.z + proximity_fade_distance, proximity_view_pos.z, VERTEX.z), 0.0, 1.0);
```

**Review by clayjohn — COMMENTED** [review]
We discussed this in the rendering meeting today and agreed that it is good to move forward. 

I just have one nitpick about variable naming, then it should be good to merge. As discussed we can leave [huwpascoe](https://github.com/huwpascoe)'s distance correction for a follow up PR

**Comment by akien-mga** [comment]
Poke @Calinou to adjust code based on review comments, then it should be good to merge.

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**